### ISSUE TITLE
[ppc] opencl compile warnings (Trac #1516)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1508]
Needs fixing:

```text
/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx: In
    constructor ‘xppc::gpu::gpu(int)’:
/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:162:59:
    warning: passing NULL to non-pointer argument 3 of ‘cl_int
    clGetDeviceInfo(cl_device_id, cl_device_info, size_t, void *, size_t *)’
    [-Wconversion-null]
clGetDeviceInfo(devID, CL_DEVICE_VENDOR, NULL, NULL, &siz);
^
```

Need something like `#define CL_USE_DEPRECATED_OPENCL_1_1_APIS` to get rid of the below warnings.

```text
/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:190:7:
    warning: ‘_cl_command_queue * clCreateCommandQueue(
        cl_context, cl_device_id, cl_command_queue_properties, cl_int *)’ is
    deprecated [-Wdeprecated-declarations]
cq = clCreateCommandQueue(ctx, devID, CL_QUEUE_PROFILING_ENABLE, &err);
    checkError(err);
^
In file included from /usr/include/CL/opencl.h:42:0,
from /home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:16:
/usr/include/CL/cl.h:1359:1: note: declared here
clCreateCommandQueue(
        cl_context                     /* context */,
^
/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:190:7:
    warning: ‘_cl_command_queue * clCreateCommandQueue(
        cl_context, cl_device_id, cl_command_queue_properties, cl_int *)’ is
    deprecated [-Wdeprecated-declarations]
cq = clCreateCommandQueue(ctx, devID, CL_QUEUE_PROFILING_ENABLE, &err);
    checkError(err);
^
In file included from /usr/include/CL/opencl.h:42:0,
from /home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:16:
/usr/include/CL/cl.h:1359:1: note: declared here
clCreateCommandQueue(
        cl_context                     /* context */,
^
/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:190:71:
    warning: ‘_cl_command_queue * clCreateCommandQueue(
        cl_context, cl_device_id, cl_command_queue_properties, cl_int *)’ is
    deprecated [-Wdeprecated-declarations]
cq = clCreateCommandQueue(ctx, devID, CL_QUEUE_PROFILING_ENABLE, &err);
    checkError(err);
^
In file included from /usr/include/CL/opencl.h:42:0,
from /home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:16:
/usr/include/CL/cl.h:1359:1: note: declared here
clCreateCommandQueue(
        cl_context                     /* context */,
^
/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx: In member
    function ‘void xppc::gpu::kernel_f()’:
/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:381:13:
    warning: ‘cl_int clEnqueueTask(
        cl_command_queue, cl_kernel, cl_uint, _cl_event * const *, _cl_event
        **)’ is deprecated [-Wdeprecated-declarations]
checkError(clEnqueueTask(cq, clkernel, 0, NULL, NULL));
^
In file included from /usr/include/CL/opencl.h:42:0,
from /home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:16:
/usr/include/CL/cl.h:1373:1: note: declared here
clEnqueueTask(
        cl_command_queue  /* command_queue */,
^
/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:381:13:
    warning: ‘cl_int clEnqueueTask(
        cl_command_queue, cl_kernel, cl_uint, _cl_event * const *, _cl_event
        **)’ is deprecated [-Wdeprecated-declarations]
checkError(clEnqueueTask(cq, clkernel, 0, NULL, NULL));
^
In file included from /usr/include/CL/opencl.h:42:0,
from /home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:16:
/usr/include/CL/cl.h:1373:1: note: declared here
clEnqueueTask(
        cl_command_queue  /* command_queue */,
^
/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:381:54:
    warning: ‘cl_int clEnqueueTask(
        cl_command_queue, cl_kernel, cl_uint, _cl_event * const *, _cl_event
        **)’ is deprecated [-Wdeprecated-declarations]
checkError(clEnqueueTask(cq, clkernel, 0, NULL, NULL));
^
In file included from /usr/include/CL/opencl.h:42:0,
from /home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:16:
/usr/include/CL/cl.h:1373:1: note: declared here
clEnqueueTask(
        cl_command_queue  /* command_queue */,
^
```

Migrated from https://code.icecube.wisc.edu/ticket/1516
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:13:35", 
    "description": "Needs fixing:\n\n{{{\n/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx: In\n    constructor \u2018xppc::gpu::gpu(int)\u2019:\n/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:162:59:\n    warning: passing NULL to non-pointer argument 3 of \u2018cl_int\n    clGetDeviceInfo(cl_device_id, cl_device_info, size_t, void *, size_t *)\u2019\n    [-Wconversion-null]\nclGetDeviceInfo(devID, CL_DEVICE_VENDOR, NULL, NULL, &siz);\n^\n}}}\n\nNeed something like `#define CL_USE_DEPRECATED_OPENCL_1_1_APIS` to get rid of the below warnings.\n\n{{{\n/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:190:7:\n    warning: \u2018_cl_command_queue * clCreateCommandQueue(\n        cl_context, cl_device_id, cl_command_queue_properties, cl_int *)\u2019 is\n    deprecated [-Wdeprecated-declarations]\ncq = clCreateCommandQueue(ctx, devID, CL_QUEUE_PROFILING_ENABLE, &err);\n    checkError(err);\n^\nIn file included from /usr/include/CL/opencl.h:42:0,\nfrom /home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:16:\n/usr/include/CL/cl.h:1359:1: note: declared here\nclCreateCommandQueue(\n        cl_context                     /* context */,\n^\n/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:190:7:\n    warning: \u2018_cl_command_queue * clCreateCommandQueue(\n        cl_context, cl_device_id, cl_command_queue_properties, cl_int *)\u2019 is\n    deprecated [-Wdeprecated-declarations]\ncq = clCreateCommandQueue(ctx, devID, CL_QUEUE_PROFILING_ENABLE, &err);\n    checkError(err);\n^\nIn file included from /usr/include/CL/opencl.h:42:0,\nfrom /home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:16:\n/usr/include/CL/cl.h:1359:1: note: declared here\nclCreateCommandQueue(\n        cl_context                     /* context */,\n^\n/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:190:71:\n    warning: \u2018_cl_command_queue * clCreateCommandQueue(\n        cl_context, cl_device_id, cl_command_queue_properties, cl_int *)\u2019 is\n    deprecated [-Wdeprecated-declarations]\ncq = clCreateCommandQueue(ctx, devID, CL_QUEUE_PROFILING_ENABLE, &err);\n    checkError(err);\n^\nIn file included from /usr/include/CL/opencl.h:42:0,\nfrom /home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:16:\n/usr/include/CL/cl.h:1359:1: note: declared here\nclCreateCommandQueue(\n        cl_context                     /* context */,\n^\n/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx: In member\n    function \u2018void xppc::gpu::kernel_f()\u2019:\n/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:381:13:\n    warning: \u2018cl_int clEnqueueTask(\n        cl_command_queue, cl_kernel, cl_uint, _cl_event * const *, _cl_event\n        **)\u2019 is deprecated [-Wdeprecated-declarations]\ncheckError(clEnqueueTask(cq, clkernel, 0, NULL, NULL));\n^\nIn file included from /usr/include/CL/opencl.h:42:0,\nfrom /home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:16:\n/usr/include/CL/cl.h:1373:1: note: declared here\nclEnqueueTask(\n        cl_command_queue  /* command_queue */,\n^\n/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:381:13:\n    warning: \u2018cl_int clEnqueueTask(\n        cl_command_queue, cl_kernel, cl_uint, _cl_event * const *, _cl_event\n        **)\u2019 is deprecated [-Wdeprecated-declarations]\ncheckError(clEnqueueTask(cq, clkernel, 0, NULL, NULL));\n^\nIn file included from /usr/include/CL/opencl.h:42:0,\nfrom /home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:16:\n/usr/include/CL/cl.h:1373:1: note: declared here\nclEnqueueTask(\n        cl_command_queue  /* command_queue */,\n^\n/home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:381:54:\n    warning: \u2018cl_int clEnqueueTask(\n        cl_command_queue, cl_kernel, cl_uint, _cl_event * const *, _cl_event\n        **)\u2019 is deprecated [-Wdeprecated-declarations]\ncheckError(clEnqueueTask(cq, clkernel, 0, NULL, NULL));\n^\nIn file included from /usr/include/CL/opencl.h:42:0,\nfrom /home/dschultz/Documents/combo/trunk/src/ppc/private/ppc/ocl/ppc.cxx:16:\n/usr/include/CL/cl.h:1373:1: note: declared here\nclEnqueueTask(\n        cl_command_queue  /* command_queue */,\n^\n}}}", 
    "reporter": "david.schultz", 
    "cc": "olivas", 
    "resolution": "fixed", 
    "_ts": "1550067215093672", 
    "component": "combo simulation", 
    "summary": "[ppc] opencl compile warnings", 
    "priority": "major", 
    "keywords": "", 
    "time": "2016-01-20T18:03:33", 
    "milestone": "", 
    "owner": "dima", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20160125T16:57:54: 
* icecube changed status from "new" to "closed"
* icecube commented:

> could not reproduce the problem; tried on cobalt. David said this shows on newer version of gcc, I have no way to test this.
> 
> Some of these are incorrect: the first warning complains "passing NULL to non-pointer argument". However, in the OpenCL specification this is a valid use: 
> 
>    "If param_value is NULL, it is ignored."
> 


* icecube changed resolution from "" to "worksforme"

**Comment by icecube-trac** [comment]
Trac update at 20160125T17:27:05: 
* david.schultz changed _comment0 from:

> On cobalt:
> 
> {{{
> dschultz@cobalt01 /scratch/dschultz/simulation/build_4.8 $ source /opt/rh/devtoolset-2/enable
> dschultz@cobalt01 /scratch/dschultz/simulation/build_4.8 $ make ppc
> [100%] Building CXX object ppc/CMakeFiles/xppc.dir/private/ppc/ocl/ppc.cxx.o
> /scratch/dschultz/simulation/src/ppc/private/ppc/ocl/ppc.cxx: In constructor ‘xppc::gpu::gpu(int)’:
> /scratch/dschultz/simulation/src/ppc/private/ppc/ocl/ppc.cxx:162:59: warning: passing NULL to non-pointer argument 3 of ‘cl_int clGetDeviceInfo(cl_device_id, cl_device_info, size_t, void*, size_t*)’ [-Wconversion-null]
>   clGetDeviceInfo(devID, CL_DEVICE_VENDOR, NULL, NULL, &siz);
>                                                            ^
> }}}
> 
> The third valid is "param_value_size", not "param_value".  So it can't be NULL, even by the OpenCL specification.

to:

> 1453742843055413


* david.schultz changed status from "closed" to "reopened"
* david.schultz commented:

> On cobalt:
> 
> 
> ```text
> dschultz@cobalt01 /scratch/dschultz/simulation/build_4.8 $ source /opt/rh/devtoolset-2/enable
> dschultz@cobalt01 /scratch/dschultz/simulation/build_4.8 $ make ppc
> [100%] Building CXX object ppc/CMakeFiles/xppc.dir/private/ppc/ocl/ppc.cxx.o
> /scratch/dschultz/simulation/src/ppc/private/ppc/ocl/ppc.cxx: In constructor ‘xppc::gpu::gpu(int)’:
> /scratch/dschultz/simulation/src/ppc/private/ppc/ocl/ppc.cxx:162:59: warning: passing NULL to non-pointer argument 3 of ‘cl_int clGetDeviceInfo(cl_device_id, cl_device_info, size_t, void*, size_t*)’ [-Wconversion-null]
>   clGetDeviceInfo(devID, CL_DEVICE_VENDOR, NULL, NULL, &siz);
>                                                            ^
> ```
> 
> The third value is "param_value_size", not "param_value".  So it can't be NULL, even by the OpenCL specification.


* david.schultz changed resolution from "worksforme" to ""

**Comment by icecube-trac** [comment]
Trac update at 20160125T17:31:17: kjmeagher commented:

> `param_value_size` is of type `size_t` so it really shouldn't be NULL

**Comment by icecube-trac** [comment]
Trac update at 20160125T18:00:15: icecube commented:

> Here is a reference that validates such use: https://www.khronos.org/registry/cl/sdk/1.2/docs/man/xhtml/clGetDeviceInfo.html. Maybe there is a newer specification that I didn't find?

**Comment by icecube-trac** [comment]
Trac update at 20160125T18:05:10: kjmeagher commented:

> Re-read that document, nowhere does is say you should pass `NULL` to `param_value_size`. It says you can pass `NULL` to `param_value` and `param_value_size_ret`, but nowhere does it say you should pass it to `param_value_size`

**Comment by icecube-trac** [comment]
Trac update at 20160125T18:13:17: icecube commented:

> you guys are right. Sorry about the noise. Change committed

**Comment by icecube-trac** [comment]
Trac update at 20160125T19:14:33: 
* david.schultz changed status from "reopened" to "closed"
* david.schultz commented:

> In changeset 141228:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="141228"
> really get rid of the opencl deprecation warnings. closes #1508
> ```


* david.schultz changed resolution from "" to "fixed"

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:13:35: 
* olivas commented:

> Milestone IceSim 5.1 deleted


* olivas changed milestone from "IceSim 5.1" to ""
